### PR TITLE
Us 20

### PR DESCRIPTION
**Pull request by MaaaChu**
## Description
Please explain briefly the changes you made here.

## Checklist
- [x] All manual security tests passing (Described here: https://github.com/Broom94/Klattice/wiki/Manual-Security-Testing)
- [x] All unit tests passing 
- [x] All UI tests passing
- [x] Code coverage suitable
- [x] WAVE toolbar run on new or updated screens
- [x] Branch named {feature|hotfix|task}/{US-.*}
- [x] If your pull request depends on any other, please link them
- [x] Approval from Tech lead
- [x] Approval from team member
- [x] Commit messages are meaningful
- [x] Target is green https://travis-ci.com/Broom94/Klattice/jobs/216924893 
- [x] Add `DO NOT MERGE` if you want to postpone merge (in labels)

### CODE REVIEW

File: api/SHA.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+exports.saltedHash= function(password)
+{
+    alg = 'sha1'
+    salt = "S@E1F53135E559C253assdk100101"; //random salt (taken from wikipedia)
+    password += salt;
+    password = hash(alg).update(password).digest('hex');
+    console.log(password);
+    return password; 
+}*/
```
Comment (matthewdoreilly) [inline api/SHA.js]: Delete this file

File: api/app.js [hunk 2]
```diff
@@ -122,7 +125,7 @@ async function addRoleToDB(roleObject){
 
 async function authenticate(username, password, res){
     var authStatus = await db.getUser(username, password);
-
+    console.log("after database, authstatus "+ authStatus);
```
Comment (matthewdoreilly) [inline api/app.js]: Remove this log

File: api/SHA.js (hunk 1 above)
Comment (Mark-Spiers) [inline api/SHA.js]: sha 1, and no newline at end

Comment (weesteeky) [inline api/SHA.js]: THIS HAS BEEN REMOVED, CAN BE DELETED

File: api/app.js [hunk 3]
```diff
@@ -145,6 +148,26 @@ async function authenticate(username, password, res){
...
+    return didRoleAdd;
+}
+
+async function deleteRole(roleObject){
+    var didRoleDelete = -1;
+    didRoleDelete = await db.deleteRole(roleObject);
+    return didRoleDelete;
+}
+
+function saltedHash(password){
```
Comment (weesteeky) [inline api/app.js]: this can be removed

File: api/app.js [hunk 4]
```diff
@@ -145,6 +148,26 @@ async function authenticate(username, password, res){
     })
 }
 
+async function addRoleToDB(roleObject){
+    var didRoleAdd = -1;
+    didRoleAdd = await db.addRole(roleObject);
+    return didRoleAdd;
```
Comment (matthewdoreilly) [inline api/app.js]: What is `didRoleAdd` used for?

Also you can just do `return await db.addRole(roleObject);` here

File: api/db.js [hunk 5]
```diff
@@ -18,15 +18,16 @@ exports.getRoleFamilies = function (callback) {
     );
 }
 
-exports.getUser = function (username, userPassword){   
+exports.getUser = function (userName, userPassword){
```
Comment (matthewdoreilly) [inline api/db.js]: Why the capitalisation here? It should be `username`

File: api/app.js [hunk 6]
```diff
@@ -145,6 +148,26 @@ async function authenticate(username, password, res){
...
+    return didRoleDelete;
+}
+
+function saltedHash(password){
+    alg = 'sha1'
+    salt = "S@E1F53135E559C253assdk100101";
+    password += salt;
+    password = hash(alg).update(password).digest('hex');
+    return password; 
+}
```
Comment (Mark-Spiers) [inline api/app.js]: this is now another file, no need for saltedHash

File: klattice/src/app/components/admin-role/admin-role.component.html [hunk 7]
```diff
@@ -82,6 +82,7 @@ <h2 class="float-left">Modify Role</h2>
         </option>
       </select>
     </th>
+    <th></th>
```
Comment (matthewdoreilly) [inline klattice/src/app/components/admin-role/admin-role.component.html]: Empty table header?

File: api/app.js [hunk 8]
```diff
@@ -37,20 +37,23 @@ app.get("/user-details", function(req, res) {
...
-    await authenticate(username, password, res); 
+  var username = req.body.params.username;
+  var password = req.body.params.password;
+  await authenticate(username, password, res); 
 });
 
 app.post('/add-role', async function(req, res){
-  ans = await addRoleToDB(req.body);
- });
+    ans = await addRoleToDB(req.body);
```
Comment (matthewdoreilly) [inline api/app.js]: What is `ans`? And what is it used for here?

File: api/db.js [hunk 9]
```diff
@@ -41,6 +42,28 @@ exports.getUser = function (username, userPassword){
 });
 }
 
+exports.addRole = function (roleObject){
```
Comment (Mark-Spiers) [inline api/db.js]: This is a duplicate of a function already in the file

File: klattice/src/app/components/admin-role/admin-role.component.html (hunk 7 above)
Comment (MaaaChu) [reply klattice/src/app/components/admin-role/admin-role.component.html]: This is to extend the table header over the delete button

File: klattice/src/app/components/admin-role/admin-role.component.html [hunk 10]
```diff
@@ -220,55 +231,109 @@ <h4 class="modal-title" id="modal-basic-title" style="color: #ffffff">
...
-      class="btn btn-outline-dark"
-      value="Cancel"
-      (click)="cancelForm()"
-    >
-      Cancel
-    </button>
-  </div>
-</ng-template>
+  </div></ng-template
+>
```
Comment (matthewdoreilly) [inline klattice/src/app/components/admin-role/admin-role.component.html]: Fix the formatting here

File: klattice/src/app/components/admin-role/admin-role.component.html (hunk 7 above)
Comment (matthewdoreilly) [reply klattice/src/app/components/admin-role/admin-role.component.html]: Seems strange to have an empty header here. Can you use CSS here instead maybe?

File: klattice/src/app/components/login/login.component.ts [hunk 11]
```diff
@@ -54,21 +54,22 @@ export class LoginComponent {
...
           window.location.reload();
-        } else if (rsp == "e") {
+        } else if (rsp == 'e') {
           this.successfulLogin(false, username);
           window.location.reload();
-        } else if (rsp == "f") {
-          alert("Failed");
+        } else if (rsp == 'f') {
+          alert('Failed');
           window.location.reload();
```
Comment (matthewdoreilly) [inline klattice/src/app/components/login/login.component.ts]: Move `window.location.reload();` out of all these `if` statements, and call it once at the end of this `if` block

File: api/app.js [hunk 12]
```diff
@@ -37,20 +37,23 @@ app.get("/user-details", function(req, res) {
...
-    await authenticate(username, password, res); 
+  var username = req.body.params.username;
+  var password = req.body.params.password;
+  await authenticate(username, password, res); 
 });
 
 app.post('/add-role', async function(req, res){
-  ans = await addRoleToDB(req.body);
- });
+   await addRoleToDB(req.body);
```
Comment (matthewdoreilly) [inline api/app.js]: You can just do `return await db.addRole(req.body);` here, and delete the `addRoleToDB` method

File: api/app.js [hunk 13]
```diff
@@ -37,20 +37,23 @@ app.get("/user-details", function(req, res) {
...
+});
 
-app.get("/capability", function(req, res) {
-  updateCapability(function() {
-    res.send(capability);
-  });
+app.post('/delete-role', async function(req, res){
+    await deleteRole(req.body);
```
Comment (matthewdoreilly) [inline api/app.js]: Same as before. You can just do `return await db.deleteRole(req.body);` here, and delete the `deleteRole` method

**Pull request merged**